fs.go

```diff
@@ -0,0 +1,253 @@
+	return f.buffer.Read(b)
+}
+
+// ReadDir will return a []*File as an []fs.DirEntry
+// Returned entries will be sorted by filename as required by fs.ReadDir
+// If n < 0, ReadDir will return all remaining entries.
+func (f *File) ReadDir(n int) ([]fs.DirEntry, error) {
+	if len(f.children) == 0 {
+		return []fs.DirEntry{}, nil
+	}
+	start := f.childIdx
+	end := f.childIdx + n
+	if end >= len(f.children) || n < 0 {
+		end = len(f.children)
+	}
+	defer func() { f.childIdx = end }()
+	names := f.childrenSorted[start:end]
+	return f.readDir(names...)
+}
+
+// readDir takes the list of children names and returns an []fs.DirEntry in the same order.
+// Will return an fs.PathError if a child name does not exist.
+func (f *File) readDir(children ...string) ([]fs.DirEntry, error) {
+	result := make([]fs.DirEntry, 0, len(children))
+	for _, name := range children {
+		child, exists := f.children[name]
+		if !exists {
+			return []fs.DirEntry{}, &fs.PathError{Op: "readdir", Path: name, Err: fs.ErrNotExist}
+		}
+		result = append(result, child.file())
+	}
+	return result, nil
+}
+
+// Close will set the underlying buffer to nil
+// If the buffer is already nil then Close will return an fs.ErrClosed
+func (f *File) Close() error {
+	if f.buffer == nil {
+		return fs.ErrClosed
+	}
+	f.buffer = nil
+	return nil
+}
+
+// fileStat stores data as an implementation of fs.FileInfo
+type fileStat struct {
+	name string
+	size int64
+	mode fs.FileMode
+	sys  interface{}
+}
+
+// IsDir will be true if a DirEnt has children
+func (s fileStat) IsDir() bool {
+	return s.mode.IsDir()
+}
+
+// ModTime is meaningless. Testmark hunks don't have a modtime
+// ModTime will always return a time.Time zero value
+func (fileStat) ModTime() time.Time {
+	return time.Time{}
+}
+
+// Mode returns the FileMode for the file.
+// The filemode will have the fs.ModeDir bit enabled if the File contains references to children
+func (s fileStat) Mode() fs.FileMode {
+	return s.mode
+}
+
+// Name returns the base name of the file
+func (s fileStat) Name() string {
+	return s.name
+}
+
+// Size returns the size of the underlying hunk when the file was opened
+func (s fileStat) Size() int64 {
+	return s.size
+}
+
+// Sys returns the _thing_ from whence this file came.
+// Generally that's the *DirEnt
+func (s fileStat) Sys() interface{} {
+	return s.sys
+}
+
+// Open opens the named hunk or a directory path.
+// Open does not follow relative paths such as .. or .
+//
+// When Open returns an error, it should be of type *PathError
+// with the Op field set to "open", the Path field set to name,
+// and the Err field describing the problem.
+//
+// Open does NOT follow conventions for fs.ValidPath.
+// Opening an empty path will return the root directory for the document.
+// This is different than the fs.ValidPath special case of using "." as the root path.
+// The testmark document treats "." and ".." the same as any other character.
+func (doc *Document) Open(name string) (fs.File, error) {
```

Comment:
I think I'd prefer to flip this around into a `func OpenAsFile(doc *Document) (fs.File, error)`.  Slightly less coupled; I don't like object-oriented flavor in general.  Is also what would happen if we pulled the FS bits out into a subpackage.

(The function name not being important, just the first thing that came to mind.  Something shorter would probably be better, if you can think something that fits.)

Reply:
Yeah this pattern is part of the `fs.FS` interface. Moving it all to its own package has some advantages here.